include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -2310,6 +2311,194 @@ namespace internal
       }
   }
 
+
+
+  /**
+   * Struct to avoid using Tensor<1, dim, Point<dim2>> in
+   * evaluate_tensor_product_value_and_gradient because a Point cannot be used
+   * within Tensor. Instead, a specialization of this struct upcasts the point
+   * to a Tensor<1,dim>.
+   */
+  template <typename Number, typename Number2>
+  struct ProductTypeNoPoint
+  {
+    using type = typename ProductType<Number, Number2>::type;
+  };
+
+  template <int dim, typename Number, typename Number2>
+  struct ProductTypeNoPoint<Point<dim, Number>, Number2>
+  {
+    using type = typename ProductType<Tensor<1, dim, Number>, Number2>::type;
+  };
+
+
+
+  /**
+   * Compute the polynomial interpolation of a tensor product shape function
+   * $\varphi_i$ given a vector of coefficients $u_i$ in the form
+   * $u_h(\mathbf{x}) = \sum_{i=1}^{k^d} \varphi_i(\mathbf{x}) u_i$. The shape
+   * functions $\varphi_i(\mathbf{x}) =
+   * \prod_{d=1}^{\text{dim}}\varphi_{i_d}^\text{1D}(x_d)$ represent a tensor
+   * product. The function returns a pair with the value of the interpolation
+   * as the first component and the gradient in reference coordinates as the
+   * second component.
+   *
+   * @param poly The underlying one-dimensional polynomial basis
+   * $\{\varphi^{1D}_{i_1}\}$ given as a vector of polynomials.
+   *
+   * @param values The expansion coefficients $u_i$ of type `Number` in
+   * the polynomial interpolation. The coefficients can be simply `double`
+   * variables but e.g. also Point<spacedim> in case they define arithmetic
+   * operations with the type `Number2`.
+   *
+   * @param p The position in reference coordinates where the interpolation
+   * should be interpolated.
+   *
+   * @param bilinear Flag to specify whether a d-linear (linear in 1D,
+   * bi-linear in 2D, tri-linear in 3D) interpolation should be made, which
+   * allows to unroll loops and considerably speed up evaluation.
+   *
+   * @param renumber Optional parameter to specify a renumbering in the
+   * coefficient vector, assuming that `values[renumber[i]]` returns
+   * the lexicographic (tensor product) entry of the coefficients. If the
+   * vector is entry, the values are assumed to be sorted lexicographically.
+   */
+  template <int dim, typename Number, typename Number2>
+  inline std::pair<
+    typename ProductTypeNoPoint<Number, Number2>::type,
+    Tensor<1, dim, typename ProductTypeNoPoint<Number, Number2>::type>>
+  evaluate_tensor_product_value_and_gradient(
```

Comment:
This function has been moved from source/fe/mapping_q_generic.cc.

Reply:
Yes, again with adjusted types and some template magic.